src/go/configgenerator/filtergen/grpc_field_extraction.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filtergen
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util/httppattern"
+	extractionpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/grpc_field_extraction/v3"
+	glog "github.com/golang/glog"
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+	"google.golang.org/protobuf/proto"
+	protov2 "google.golang.org/protobuf/proto"
+	descpb "google.golang.org/protobuf/types/descriptorpb"
+
+	policypb "google3/google/api/policy_go_proto"
+	corepb "google3/third_party/envoy/src/api/envoy/config/core/v3/pkg_go_proto"
+)
+
+const (
+	// GRPCFieldExtractionFilterName is the debug name of the extension.
+	GRPCFieldExtractionFilterName = "envoy.filters.http.grpc_field_extraction"
+)
+
+// GRPCFieldExtractionGenerator is a filter generator for the filter config.
+type GRPCFieldExtractionGenerator struct {
+	// ProtoDescriptorBin is the serialized proto descriptor set.
+	ProtoDescriptorBin []byte
+
+	// PoliciesBySelector maps from OP selector to OSS MethodPolicy.
+	PoliciesBySelector map[string]*policypb.MethodPolicy
+
+	NoopFilterGenerator
+}
+
+// NewGRPCFieldExtractionFilterGensFromOPConfig creates a GRPCFieldExtractionGenerator from
+// OP service config + descriptor + ESPv2 options. It is a FilterGeneratorOPFactory.
+func NewGRPCFieldExtractionFilterGensFromOPConfig(serviceConfig *confpb.Service, opts options.ConfigGeneratorOptions) ([]FilterGenerator, error) {
+	grpcGen, err := NewGRPCFieldExtractionFilterGenFromOPConfig(serviceConfig, opts, true)
+	if err != nil {
+		return nil, err
+	}
+	if grpcGen == nil {
+		return nil, nil
+	}
+
+	return []FilterGenerator{
+		grpcGen,
+	}, nil
+}
+
+// NewGRPCFieldExtractionFilterGenFromOPConfig creates a single GRPCFielcExtractorGenerator.
+//
+// It also has the option to skip the filter. If enabled, all checks will occur
+// and filter may not be created. Otherwise, filter is always created.
+func NewGRPCFieldExtractionFilterGenFromOPConfig(serviceConfig *confpb.Service, opts options.ConfigGeneratorOptions, maybeSkipFilter bool) (*GRPCFieldExtractionGenerator, error) {
+	var descriptorBin []byte
```

Comment:
Similar to transcoder filter, you should check if gRPC support is required first, and skip filter if not. We don't need this filter if user configured OpenAPI backend (supported for ESPv2).

```golang
	isGRPCSupportRequired, err := IsGRPCSupportRequiredForOPConfig(serviceConfig, opts)
	if err != nil {
		return nil, err
	}
	if !isGRPCSupportRequired {
		glog.Infof("gRPC support is NOT required, skip gRPC field extraction filter completely.")
		return nil, nil
	}
```